### ISSUE TITLE
Out of memory handling in MQTTAsync client

### ISSUE THREAD

**Issue by jpwsutton** [eclipse-paho/paho.mqtt.c#45]
migrated from Bugzilla [#444229](https://bugs.eclipse.org/bugs/show_bug.cgi?id=444229)
status ASSIGNED severity _normal_ in component _MQTT-C_ for _1.2_
Reported in version _1.0_ on platform _PC_
Assigned to: Ian Craggs

On 2014-09-16 06:20:32 -0400, Ian Craggs wrote:

> 2) NULL checks
> There are no NULL checks on malloc. The current implementation will end up with segmentation faults on low memory situations.
> I assume this is not uncommon on small devices.
> 
> > 2) NULL checks for malloc are handled in Heap.c.  The malloc call is
> > overridden.  This will result in error messages in the trace. 
> 
> Looking into Heap.c reveals that this code is aware of OOM situations and reacts appropriately.
> So Heap.c may return NULL to the MQTTAsync.c where no check happens.
> 
> > However,
> > the question is what should the action be in the case of malloc
> > failure?   I guess the best solution is that the API call returns with
> > failure - memory error.  
> 
> Agreed on API level.
> For code running in background threads it will be more complicated since it has potentially to report the OOM situation to the API consumer.
> And if this paths also require memory then they need to be changed.
> 
> > However that would add considerable complexity
> > and code for handling the error and cleaning up partially allocated
> > memory structures.  
> > Yes. OOM handling is IMO part of the essential complexity of any C/C++ component.

On 2014-09-16 06:30:00 -0400, Ian Craggs wrote:

> Out of memory handling is "essential".  It's also a balance of risk vs implementation complexity - that is, what is it worth?  
> 
> I think that I became less motivated to handle these situations after seeing how in general Java applications cope with the same situation - usually badly.
> 
> However, I think there are some relatively high risk situtations here which we should cope with.  While there are small data structures allocated, the biggest risk comes from topic strings and MQTT payloads, which have a maximum size of 64k and 256MB respectively.  So I would suggest we cater for those situations first, and assess whether the rest are needed.  
> 
> There is an MQTT problem with the receipt of messages larger than we can handle - if the message is QoS 1 or 2 and we close the connection because we cannot receive the message, then the server will send it again.  We could receive the entire message piecemeal and then pass it incomplete to the application, but that would violate the spirit of QoS 1 or 2.  There is no clean solution to this, save changing the protocol
> 
> (For all the small allocations which could fail, I would rather implement one global strategy rather than every individual check.  Using setjmp/longjmp comes to mind.  But these are relatively low risk, except perhaps if we have a lot of buffered messages.  The amount of buffering can be controlled by the application however.)

On 2014-09-17 08:35:33 -0400, Juergen Denner wrote:

> I think the goals could be:
> 
> 1) Stability
>    a) Never segfault in any case and inform the consumer.
>    b) Reach a deterministic state and try to recover.
>       If recover is not possible the shutdown safely and refuse new API calls
>       with a clear error message.
> 2) Handle OOM 'appropriately' according to the MQTT protocol for oversized
>    messages.
> 
> I see the following approaches on handling oversized payloads (oversized is determined by the execution context). 
> 1. Reject or skip the message - this seems not to be supported by the MQTT spec.
>    Is this a gap?
> 2. What you name 'piecemeal' can be named streaming. 
>    Streaming is a pretty common strategy in handling arbitrary data size.
>    So the message could be either 
>    a) Send partially to the receiver including a marker for the last segment.
>       (data push). No trivial to get it right including error handling.
> - or -
>   b) First received to the persistence and then it could be consumed by the
>     client, without the risk on blocking due to network (data pull).
>   However, I don't see the relevance for this approach on small devices.
> 1. Consume and 'drop' the data, inform the client that an oversized message was
>    dropped. This would at least unblock the client.
> 
> (Java related comments)
> Java never segfaults on OOM (at least due to Java heap ; an empty C heap do crash some JVM types). Java has a clear OOM strategy where you normally never react directly on OOM. If you take care for (runtime) exception and do a proper resource cleanup then you are normally safe. Some containers recover from such situations â most donât.

On 2014-09-17 11:11:19 -0400, Ian Craggs wrote:

> Java might have a clear out of memory strategy, but whenever I've tried to recover from those situations in real tests, it never worked.  This was some time ago, so things might have changed.  Anyway that's a digression.
> 
> I think a goal should also be that the fixes ought to be testable and cater for likely to happen scenarios.   Note that the Linux man page for malloc says:  
> 
> ```
>    By  default,  Linux  follows an optimistic memory allocation strategy.  This means that when
>    malloc() returns non-NULL there is no guarantee that the memory really is  available.   This
>    is  a  really  bad  bug.  In case it turns out that the system is out of memory, one or more
>    processes will be killed by the infamous OOM killer.  In case Linux is employed  under  circumstances
>    where it would be less desirable to suddenly lose some randomly picked processes,
>    and moreover the kernel version is sufficiently recent, one can switch off this  overcommitting
>    behavior using a command like:
> 
>        # echo 2 > /proc/sys/vm/overcommit_memory
> 
>    See   also   the   kernel   Documentation   directory,  files  vm/overcommit-accounting  and
>    sysctl/vm.txt.
> ```
> 
> So, we can't guarantee freedom from segfaults.
> 
> Goals:
> 
> 1) 
> a) impossible in the face of OS bugs.   For allocation failures internal to this API, ideally yes. 
> b) I think trying to recover is too much work for out of memory conditions for this API.  Unless by recovering you mean removing all the library state and requiring the application to start again.
> 
> 2) "Oversized" MQTT messages.  Two cases: incoming and outgoing.
> 
> Outgoing: just return failure to the application.  Most likely to happen on topic or payload allocation.  There could be a max message size config parameter.
> 
> Incoming: 
> 1) rejecting the message is not supported by MQTT.  This is a well known requirement for improvement.
> Skipping the message is not allowed for QoS 1 or 2.  It is allowed for QoS 0.
> 
> 2) I think streaming or fragmentation is over complicated for the goals of this API.  
> 
> 3) not strictly allowed by the MQTT specification.
> 
> The only strictly allowed action for a message to big to be received when the QoS > 0, is to close the session.  But also in the case of QoS > 0, this will result in the message being sent again, unless we have connected or will connect with clean session = true.
> 
> Given the likelihood of occurrence vs cost of implementation I suggest this approach:
> 
> 1) deal with incoming over-sized topics/payloads by closing the connection when memory allocation failures are detected, with an appropriate reason code to connectionLost.
> 
> 2) deal with outgoing "over-sized" topics/payloads by returning memory allocation error to the application. 
> 
> 3) For any other situations, allocating small data structures in the depths of a function, I think cleaning up the API state entirely and requiring the application to start again is the best option.  All memory can be freed using the heap tracking routines.  But I rate this as a relatively low priority because of the low likelihood of occurrence, and low likelihood of recovery.

On 2014-09-24 02:51:37 -0400, Juergen Denner wrote:

> Yes, malloc on Linux has a really wired behavior. However - malloc for windows returns NULL in OOM situations. At least for this platform segfaults can be avoided.
> 
> Comments for goals:
> 1b)
> A valid way for me is to free the paho client and create a new one. However it must be detectable by the error code.
> 2) Many thanks for detailing this out
> On way of reducing the used memory for outgoing messages could be that the implementation does not create a copy of the topic & payload (maybe by an additional publish/send API)
> If the consumer creates the message structure using the malloc functions from the Paho client then the contract would be to hand over the memory ownership to the PahoClient which releases the memory (payload, topic name) after processing.
> 
> Comments on your suggested approach:
> 1 & 2 Yes
> 3) > requiring the application to start again â¦
> Do you mean destroying and recreating the PahoClient or do you mean the whole process that is using the Paho client?

On 2015-01-19 11:52:38 -0500, Ian Craggs wrote:

> Moving to later milestone

**Comment by LeoonLee** [comment]
I'm sorry to not reply until today,and please don't joke my Chinglis.

Let 's review the code , file:MQTTAsync.c, function:MQTTAsync_restoreCommand,

```c
MQTTAsync_command* command = NULL;

```
> This line, define the pointer named 'command', used for point to the commands which read from archive.

```c
qcommand = malloc(sizeof(MQTTAsync_queuedCommand));
memset(qcommand, '\0', sizeof(MQTTAsync_queuedCommand));
command = &qcommand->command;
```

> These code, malloc and init a MQTTAsync_queuedCommand struct; 'command' point to a MQTTAsync_command struct

MQTTAsync_command:
```c
typedef struct
{
	int type;
	MQTTAsync_onSuccess* onSuccess;
	MQTTAsync_onFailure* onFailure;
	MQTTAsync_token token;
	void* context;
	START_TIME_TYPE start_time;
	union
	{
		struct
		{
			int count;
			char** topics; //
			int* qoss; //
		} sub;
		struct
		{
			int count;
			char** topics; //
		} unsub;
		struct
		{
			char* destinationName;
			int payloadlen;
			void* payload;
			int qos;
			int retained;
		} pub;
		struct
		{
			int internal;
			int timeout;
		} dis;
		struct
		{
			int currentURI;
			int MQTTVersion; /**< current MQTT version being used to connect */
		} conn;
	} details;
} MQTTAsync_command;
```
Focus here,'topics' is a pointer point to **char***,
```
struct
		{
			int count;
			char** topics; //
			int* qoss; //
		} sub;
```

so, 'command->details.sub.topics' point to NULL,
but as a 'char **', 'command->details.sub.topics' need to malloc again before used.

Next is my code after modify:

```
static MQTTAsync_queuedCommand* MQTTAsync_restoreCommand(char* buffer, int buflen)
{
	MQTTAsync_command* command = NULL;
	MQTTAsync_queuedCommand* qcommand = NULL;
	char* ptr = buffer;
	int i;
	size_t data_size;

	FUNC_ENTRY;
	qcommand = malloc(sizeof(MQTTAsync_queuedCommand));
	memset(qcommand, '\0', sizeof(MQTTAsync_queuedCommand));
	command = &qcommand->command;

	command->type = *(int*)ptr;
	ptr += sizeof(int);

	command->token = *(MQTTAsync_token*)ptr;
	ptr += sizeof(MQTTAsync_token);

	switch (command->type)
	{
		case SUBSCRIBE:
			command->details.sub.count = *(int*)ptr;
			ptr += sizeof(int);

			if (command->details.sub.count > 0)
			{
				command->details.sub.topics = (char **)malloc(command->details.sub.count);
				command->details.sub.qoss = (int *)malloc(command->details.sub.count);
			}

			for (i = 0; i < command->details.sub.count; ++i)
			{
				data_size = strlen(ptr) + 1;

				command->details.sub.topics[i] = malloc(data_size);
				strcpy(command->details.sub.topics[i], ptr);
				ptr += data_size;

				command->details.sub.qoss[i] = *(int*)ptr;
				ptr += sizeof(int);
			}
			break;

		case UNSUBSCRIBE:
			command->details.unsub.count = *(int*)ptr;
			ptr += sizeof(int);

			if (command->details.unsub.count > 0)
			{
				command->details.unsub.topics = (char **)malloc(command->details.unsub.count);				
			}

			for (i = 0; i < command->details.unsub.count; ++i)
			{
				size_t data_size = strlen(ptr) + 1;

				command->details.unsub.topics[i] = malloc(data_size);
				strcpy(command->details.unsub.topics[i], ptr);
				ptr += data_size;
			}
			break;

		case PUBLISH:
			data_size = strlen(ptr) + 1;
			command->details.pub.destinationName = malloc(data_size);
			strcpy(command->details.pub.destinationName, ptr);
			ptr += data_size;

			command->details.pub.payloadlen = *(int*)ptr;
			ptr += sizeof(int);

			data_size = command->details.pub.payloadlen;
			command->details.pub.payload = malloc(data_size);
			memcpy(command->details.pub.payload, ptr, data_size);
			ptr += data_size;

			command->details.pub.qos = *(int*)ptr;
			ptr += sizeof(int);

			command->details.pub.retained = *(int*)ptr;
			ptr += sizeof(int);
			break;

		default:
			free(qcommand);
			qcommand = NULL;

	}

	FUNC_EXIT;
	return qcommand;
}
```

thanks.

Attach an file , it 's content a 'SUBSCRIBE' command wait for restore, you could use to debug the program step by step.


[c-1.zip](https://github.com/eclipse/paho.mqtt.c/files/1500467/c-1.zip)

**Comment by icraggs** [comment]
Superseded by #552
